### DISCUSSION TITLE
[Proposal] Restructure Essentials folder and files

### DISCUSSION THREAD

**Discussion by Andre601** [EssentialsX/Essentials#4033] [Ideas and Feature Suggestions]
## The Problem
The current config.yml of Essentials is "bloated" to put it simply.
There are tons of settings and comments from which some may not even be used or required by everyone, since not everyone needs modules like XMPP, AntiBuild, etc.

## The Idea
The idea would be to split the config up into different, dedicated config files. The inspiration for this came from looking at the [EssentialsXDiscord module][pull] which has its own dedicated folder and file to configure.

While giving each module their own folder in the plugin directory would be an idea do I feel like it would be better to have them inside the Essentials directory as their own dedicated folder, to avoid unnecessary bloat of the plugins folder.
Each module would have their own directory with a config file (i.e. EssentialsXChat would have `/plugins/Essentials/EssentialsChat/config.yml`) and perhaps other files too if they're required for the module.

The core features available in base Essentials should however stay as default config.yml in the Essentials folder.

### General Structure Idea
Assuming all modules (including Discord one) being used should you imagine the following folder structure:
```
plugins/
  |
  |- Essentials/
  |    |
  |    |- config.yml # Core features of Essentials itself
  |    |- tpr.yml # Maybe move to warps folder?
  |    |- kits.yml
  |    |- upgrades-done.yml
  |    |
  |    |- EssentialsChat/
  |    |    |
  |    |    |- config.yml # All settings of EssentialsChat
  |    |
  |    |- EssentialsSpawn/
  |    |    |
  |    |    |- config.yml # All settings of EssentialsSpawn
  |    |    |- spawn.yml # Spawn settings when created
  |    |
  |    |- EssentialsAntiBuild/
  |    |    |
  |    |    |- config.yml # All settings of EssentialsAntiBuild
  |    |
  |    |- EssentialsGeo/
  |    |    |
  |    |    |- config.yml # All settings of EssentialsGeo
  |    |
  |    |- EssentialsProtect/
  |    |    |
  |    |    |- config.yml # All settings of EssentialsProtect
  |    |
  |    |- EssentialsXMPP/
  |    |    |
  |    |    |- config.yml # All settings of EssentialsXMPP
  |    |
  |    |- EssentialsDiscord/ # Move Discord module from dedicated folder to here instead.
  |    |    |
  |    |    |- config.yml # All settings of EssentialsDiscord
  |    |
  |    |- files/
  |    |    |
  |    |    |- # Any TXT files created from commands like /rules or /motd
  |    |
  |    |- userdata/
  |    |    |
  |    |    |- # All the JSON files
  |    |    |- usermap.csv # Move the usermap file to here
  |    |
  |    |- warps/
  |    |    |
  |    |    |- # Same as before
  |    |
  |    |- items/
  |         |
  |         |- custom_items.yml
  |         |- items.json
  |         |- worth.yml
  |
  |- # All the other plugins
```

## Backwards compatibility
To not break support with old configs, and forcing people to recreate stuff all over again, would I propose the idea to make Essentials run in a "backwards-compatibility" mode for the various settings and to only prioritize the new folder and file if the folder exists.

For example: Just updating Essentials won't move/migrate settings. Only when I f.e. create the `EssentialsChat` folder will Essentials create and use the Chat-specific config file there...

Alternative (and if that is an option with the upcoming [move to configurate][configurate]) could Essentials perhaps attempt to migrate settings to the folders, if possible.

[pull]: https://github.com/EssentialsX/Essentials/pull/3844
[configurate]: https://github.com/EssentialsX/Essentials/pull/4072

**Comment by Andre601** [comment]
Just would like to get a little info of if this will be considered for a future major update, even if it was done differently.
Like I see the labels, but does that mean anything for a possible implementation/realisation of this idea?

Would love to hear what possible issues here are.

**Reply by JRoy** [reply]
> Just would like to get a little info of if this will be considered for a future major update, even if it was done differently.
> Like I see the labels, but does that mean anything for a possible implementation/realisation of this idea?
> 
> Would love to hear what possible issues here are.

I have absolutely no issues with this and would love to have it in our 2.20.0 release. The only thing I'm waiting on in terms of implementing this is for Configurate to finish up their comments support (https://github.com/SpongePowered/Configurate/pull/175) as currently editing any yml document would cause it to lose their comments.
